sympy/solvers/ode.py

```diff
@@ -5926,19 +5927,22 @@ def _remove_coefficient(expr, x):
         # This function DOES NOT do that (it doesn't even look at the
         # homogeneous equation).
         temp_set = set([])
+        temp_set_2 = set([])
         f = Function('f')
         if expr.is_Add:
             for i in expr.args:
-                act = list(_get_trial_set(i,x))
-                while any(sub_func_doit(eq_homogenous, f(x), ts).is_zero for ts in act):
-                    act = {x*ts for ts in act}
-                temp_set.union(set(act))
+                act = _get_trial_set(i,x)
+                if eq_homogenous is not S.Zero:
+                    while any(sub_func_doit(eq_homogenous, f(x), ts).is_zero for ts in act):
+                        act = {x*ts for ts in act}
```

Comment:
@oscarbenjamin in the example `eq = -x - f(x) + Derivative(f(x), x) + 5*Derivative(f(x), (x, 2)) + 3*Derivative(f(x), (x, 3)) - exp(-x)`  while statement is executed only once for `exp(-x)` whereas according to our approach it should have been executed twice as` x*exp(-x)` is also part of complimentary solutions. can you help?

Reply:
What does `sub_func_doit` return?

Reply:
It evaluates the equation by replacing f(x) with new function and it return 0 or None according to evaluation

Reply:
I mean what does it return in this particular example?

Reply:
Actual trialset which should be returned is `{1, x, x**2*exp(-x)}`   but it returns `{1, x, x*exp(-x)}`  and about that while loop for `exp(-x) ` the expression is 0 so the while loop executes and for `x*exp(-x)` it return `NONE`

Reply:
No `is_zero` gives `None`. What does `sub_func_doit` return?

Reply:
`sub_func_doit` returns `NONE`

Reply:
`[-2*x*exp(-x) + 3*(3 - x)*exp(-x) + 5*(x - 2)*exp(-x) + exp(-x)]`  it returns this if we simplify it then the ans is zero